### PR TITLE
Parser adjustments and reorganization of functional settings in stringtables

### PR DESCRIPTION
**Pull request by Dilvish-fo**
For a long time now, something like this has been on my mind because of the complications that functional stringtable entries cause for our translators (and then sometimes from their resulting mistakes).  Also, in PR #213 @geoffthemedio  and @Vezzra had concerns about the settings being stored in config.xml, and Vezzra suggested that the settings for it should instead be in a content-directory global_settings.txt, and so...
- made parse::file_substitution() less promiscuous in it's search for includes; specifically, prior to this they could start anywhere in a line and therefor could not be commented-out.  This requires that they be at the start of a line and preceded by a single '#', as in
  # include "shared_macros.txt"
- made Stringtable use parse::read_file instead of it's own identical version, and also use parse::file_substitution, so that the functional, non-translatable stringtable entries can be moved out of the stringtables.

Removed functional, non-translatable stringtable entries from being directly present in the stringtables, by incorporating them via the newly supported #include functionality for stringtables. The purpose is to minimize translator confusion and risk from translator mistake and to facilitate customization.  The functional entries were divided into three groups.
- The first group is in default/global_settings.txt.  This will be for more general settings & it may be useful in the future to have the server specify them for the clients.  This contains no currently functional setting entries, but has placeholders for the production queue frontloading and topping up factors.
- The second group is in default/content_specific_parameters.txt.  This is primarily for content specific entries that are used to determine various aspects of the UI, such as which tags get used for the Census popup, and which specials count as Growth specials for the purposes of the ObjectListWnd. If content customizations are made to these aspects of content then the player will likely want to make corresponding changes to these entries.
- The third group is in default/customizations/common_user_customizations.txt.  This is for again for entries that affect the UI, but unlike the content-specific settings, these are more a matter of personal preference, and the players is somewhat encouraged to customize these to their preference.  THe current sole entry is that controlling the order of Sitrep presentation in the Sitrep panel.

### CODE REVIEW

**Comment by geoffthemedio** [comment]
I don't think global_settings.txt makes sense, at least long-term. It sounds, more or less, like a list of game rules tweaks / options. These should be set during game setup, and sent to the server like the galaxy setup settings, and the defaults for these options in the UI should be similarly stored in the OptionsDB.

An addition to content_specific_parameters.txt (that requires some C++ support) might be a list of what buildings are considered to be shipyards for purposes of adjusting the UI display of system names. This would resolve the note in SidePanel.cpp about "This isn't the stone age anymore" regarding checking specifically for the shipyard building name in the C++ code.

File: default/global_settings.txt [hunk 1]
```diff
@@ -0,0 +1,6 @@
+
+PRODUCTION_QUEUE_FRONTLOAD_FACTOR
```
Comment (geoffthemedio) [inline default/global_settings.txt]: should probably put some comments with these to explain their effect.

File: default/species.txt [hunk 2]
```diff
@@ -1,4 +1,4 @@
-include "shared_macros.txt"
```
Comment (geoffthemedio) [inline default/species.txt:1]: wasn't it preferred to put the includes at the end of the file to preserve the line numbering for parse error output?

File: default/stringtables/en.txt [hunk 3]
```diff
@@ -14654,3 +14556,10 @@ MONSTER_NAMES
 [[SHIP_NAME_BASQUE_MYTHOLOGY]]
 [[SHIP_NAME_UNSORTED_MYTHOLOGY]]
 [[SHIP_NAME_DEVS]]'''
+
+
+
+#include "../global_settings.txt"
```
Comment (geoffthemedio) [inline default/stringtables/en.txt:14565]: (like this...)

**Comment by adrianbroher** [comment]
> Removed functional, non-translatable stringtable entries from being directly present in the stringtables, by incorporating them via the newly supported #include functionality for stringtables.

If they are not translatable in the first place the never belonged in string tables anyway. So the correct way to fix this is to create a new configuration file or adapt an existing configuration file accordingly and not adding a new include syntax to the already abysmal string tables.

File: default/species.txt (hunk 2 above)
Comment (Dilvish-fo) [inline default/species.txt:1]: makes sense, and sounds familiar, yes.  I'll do that.

File: default/global_settings.txt (hunk 1 above)
Comment (Dilvish-fo) [inline default/global_settings.txt]: Your plan to move away from using global_settings.txt certainly sounds fine; I just don't think we need to jump all the way to the endpoint in one leap (and I don't have time to get us all the way there in the immediate future).  I'll take a look at adding a shipyards list for this system name code you mention.

**Comment by Dilvish-fo** [comment]
@adrianbroher Please note that this was already an add-on from a different task.  For such things I tend to take the shortest reasonable path that is a clear improvement accomplishing my goals without significant drawbacks.  This may not be your preferred solution, but it is a reasonable path that accomplishes my goals and appears to me to be a clear improvement.  The most significant drawback I can think of is that we should probably preface these functional keys with "FUNCTIONAL_" and then include a warning at the top of the stringtable that in order to avoid potential key conflicts with functional entries, no stringtable keys should begin with "FUNCTIONAL_".

I suppose it would be a fairly moderate additional step to make new ParameterString(keyword) and a ParameterStringList(keyword) functions based off the the UserString ones but reading from a parameters.txt file (which would itself incorporate these same subfiles) instead of a stringtable.  Were you trying to make an encouraging suggestion of adding that further improvement to this initial improvement?  I am rather tight on time no, though, so I don't want to make commitments there is no compelling reason for.  It appears to me that this would be fine as a later extension of this work; stopping here for now creates no problems for the later extension.

File: default/global_settings.txt (hunk 1 above)
Comment (geoffthemedio) [inline default/global_settings.txt]: It's in SidePanel.cpp. Search for "This isn't the stone age anymore"

**Comment by adrianbroher** [comment]
> This may not be your preferred solution […]
> Were you trying to make an encouraging suggestion of adding that further improvement to this initial improvement? 

This has nothing to do with my preferred solution. I'm pointing out that the concept is broken. You already wrote why it is broken and I gave my opinion on how it should be solved, to quote myself, 'the correct way'.

> […] that functional stringtable entries cause for our translators (and then sometimes from their resulting mistakes).

They just don't belong there.
- Functional entries are an entire different thing than translations. Even the comments point out that they should not be touched by the translators. This indicates that they should not be inside the translation file at all.
- Functional entries are sorted sets of strings most of the time. The std::string UserString(const std::string& key) is not suited for accessing them in this way. This indicates that there should be an dedicated API for accessing them.
- Using the UserString call indicates: "Here is something to translate", which is not true for functional entires. It's making automatic key extraction hard than it already is. Using a dedicated API for functional entries or integrating the search order in the 'whatever needs to be ordered' declarations would solve this.
- Adding an include mechanism makes any conversion to a 'sane translation system'™ already harder than it is.

> For such things I tend to take the shortest reasonable path that is a clear improvement accomplishing my goals without significant drawbacks.

So this is a stopgap solution?

**Comment by Dilvish-fo** [comment]
@adrianbroher  Everyone else who has commented appears to consider this to be a solid improvement.   If you notice from Geoff's comment, this also allows a reasonable solution to another coding problem that had been vexing him, related to UI customization.

You appear to be succumbing to what in English is referred to as the mistake of "Letting the Perfect be the Enemy of the Good".  You point out a number of flaws with the current system, at least many of which my changes significantly improve.  I haven't seen you point out any ways my changes make anything worse; you claim the include feature does, but that would be so trivial to drop when no longer needed that I think your point is no point at all.  Your critique does not appear to be an attempt to set forth helpful suggestions for anyone later seeking to extend the work; you appear unwilling to let things stand if I don't solve every problem you see with the current approach.

You have a strong idea about what you think the end goal should be.  I feel no need to get into any significant debate about whether or not that is indeed the correct end goal, but when I pointed out that this work adds no obstacles to reaching the goal you identify, as a later extension that someone could do (indeed, I think this work makes achieving your goal easier), and that I intend to stop here, you seem unable to accept that, and respond with a lot of further critique of aspects of the current system which I leave unchanged, and appear to object to me leaving this work as-is:

> So this is a stopgap solution?

If you want to make further improvements, be my guest.

In the meantime, I believe I have addressed everyone else's concerns/requests, and I want to be able to make use of this in further UI improvements, so I'll be merging it now.

**Comment by adrianbroher** [comment]
> If you want to make further improvements, be my guest.

I will have to, because this PR is ~~stepping on my toes~~ getting in the way for another change I'm working on.

**Comment by Vezzra** [comment]
> I don't think global_settings.txt makes sense, at least long-term. It sounds, more or less, like a list of game rules tweaks / options. These should be set during game setup, and sent to the server like the galaxy setup settings, and the defaults for these options in the UI should be similarly stored in the OptionsDB.

Well yes, of course in the end all those settings should be somehow accessible through the UI during game setup. Until then however this global_settings.txt file is a good temporary solution I think, certainly better that storing game rule settings in config.xml :wink:

**Pull request merged**